File: frontend/src/api/insights.ts
/**
 * 瓶颈与产能洞察端点的类型化客户端（design.md Components §6 `/insights`，任务 13.5，R15）。
 *
 * 形状逐字对应后端 `app/api/insights.py`。全部数值确定性、无 LLM；「+20% 工时的拖期变化量」由
 * 后端沙箱实算（R15.2）。只读端点。
 */

import { apiFetch } from './client';

/** 一台机器在当前 ACTIVE 计划中的产能画像（R15.1–R15.3）。 */
export interface MachineInsight {
  readonly machine_id: string;
  readonly machine_type: string;
  readonly capabilities: readonly string[];
  readonly utilisation: number;
  readonly busy_minutes: number;
  readonly available_minutes: number;
  readonly job_count: number;
  readonly order_value_share: number;
  readonly is_critical: boolean;
  /** 可用工时 +20% 时 total_tardiness_minutes 的变化量（R15.2，沙箱实算，负=改善）。 */
  readonly tardiness_delta_if_plus_20pct: number;
}

/** 按 required_worker_skill 聚合的技能缺口（R15.4）。gap>0 表示供不应求。 */
export interface SkillGap {
  readonly skill: string;
  readonly required_minutes: number;
  readonly available_minutes: number;
  readonly gap_minutes: number;
}

export interface BottleneckInsights {
  readonly active_plan_id: string | null;
  readonly machines: readonly MachineInsight[];
  readonly skill_gaps: readonly SkillGap[];
}

/** 拉取当前 ACTIVE 计划的瓶颈与产能洞察。只读。无 ACTIVE 计划时后端返回 409 NO_ACTIVE_PLAN。 */
export function getBottlenecks(): Promise<BottleneckInsights> {
  return apiFetch<BottleneckInsights>('/insights/bottlenecks');
}

File: backend/app/api/insights.py
"""瓶颈与产能洞察端点（design.md Components §3.7 / §6 `/insights`，任务 13.5，R15）。

- `GET /api/insights/bottlenecks` —— 当前 `ACTIVE` 计划的每机器利用率 / 作业数 / 订单价值占比、
  关键机器标识、按技能聚合的缺口，以及每台机器「可用工时 +20% 时 `total_tardiness_minutes` 的
  变化量」（由 `run_capacity_sandbox(purpose=BOTTLENECK)` 实算，R15.2）。

只读端点（无需认证，与其余 GET 同口径）。全部数值确定性、无 LLM；+20% 变化量在沙箱隔离下实算，
不写任何生产数据。无 `ACTIVE` 计划 → `NO_ACTIVE_PLAN`。
"""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, ConfigDict
from sqlalchemy.orm import Session, sessionmaker

from app.api.errors import ErrorCode, NextAction, error_response
from app.services.insights import NoActivePlanError, bottleneck_insights
from app.services.runtime_clock import operational_now

router = APIRouter(prefix="/insights", tags=["insights"])


class MachineInsightOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    machine_id: str
    machine_type: str
    capabilities: list[str]
    utilisation: float
    busy_minutes: int
    available_minutes: int
    job_count: int
    order_value_share: float
    is_critical: bool
    tardiness_delta_if_plus_20pct: int


class SkillGapOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    skill: str
    required_minutes: int
    available_minutes: int
    gap_minutes: int


class BottleneckInsightsOut(BaseModel):
    model_config = ConfigDict(extra="forbid")

    active_plan_id: str | None
    machines: list[MachineInsightOut]
    skill_gaps: list[SkillGapOut]


@router.get(
    "/bottlenecks",
    response_model=BottleneckInsightsOut,
    summary="瓶颈与产能洞察：利用率 / 关键机器 / 技能缺口 / +20% 工时拖期变化（R15）",
)
def get_bottlenecks(request: Request) -> BottleneckInsightsOut | JSONResponse:
    """计算当前 `ACTIVE` 计划的瓶颈与产能洞察。只读，无 LLM。"""
    factory: sessionmaker[Session] = request.app.state.session_factory
    with factory() as db:
        try:
            result = bottleneck_insights(
                db, now=operational_now(request.app.state.settings.app_env)
            )
        except NoActivePlanError:
            return error_response(
                status_code=409,
                code=ErrorCode.NO_ACTIVE_PLAN,
                message="There is no ACTIVE plan, so bottleneck insights cannot be computed. Please generate and approve a plan first.",
                next_actions=[NextAction(action="generate_plan", href="/plans/generate")],
            )
    return BottleneckInsightsOut(
        active_plan_id=result.active_plan_id,
        machines=[
            MachineInsightOut(
                machine_id=m.machine_id,
                machine_type=m.machine_type,
                capabilities=list(m.capabilities),
                utilisation=m.utilisation,
                busy_minutes=m.busy_minutes,
                available_minutes=m.available_minutes,
                job_count=m.job_count,
                order_value_share=m.order_value_share,
                is_critical=m.is_critical,
                tardiness_delta_if_plus_20pct=m.tardiness_delta_if_plus_20pct,
            )
            for m in result.machines
        ],
        skill_gaps=[
            SkillGapOut(
                skill=g.skill,
                required_minutes=g.required_minutes,
                available_minutes=g.available_minutes,
                gap_minutes=g.gap_minutes,
            )
            for g in result.skill_gaps
        ],
    )
